File: src/main/java/kr/hahaha98757/zombiesaddon/hudposition/DelayedTask.java
package kr.hahaha98757.zombiesaddon.hudposition;

import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;

public class DelayedTask {
    private int counter;
    private final Runnable run;

    public DelayedTask(Runnable run) {
        this(run, 2);
    }

    public DelayedTask(Runnable run, int ticks) {
        this.counter = ticks;
        this.run = run;
        MinecraftForge.EVENT_BUS.register(this);
    }

    @SubscribeEvent
    public void onTick(TickEvent.ClientTickEvent event) {
        if (event.phase != TickEvent.Phase.START) return;

        if (counter <= 0) {
            MinecraftForge.EVENT_BUS.unregister(this);
            run.run();
        }

        --counter;
    }
}

File: src/main/kotlin/kr/hahaha98757/zombiesaddon/utils/DelayedTask.kt
package kr.hahaha98757.zombiesaddon.utils

import kr.hahaha98757.zombiesaddon.events.LastClientTickEvent
import kr.hahaha98757.zombiesaddon.events.ServerTickEvent
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent

class DelayedTask(private var tick: Int = 2, private val serverTick: Boolean = false, private val block: () -> Unit) {
    init {
        MinecraftForge.EVENT_BUS.register(this)
    }

    @Suppress("unused")
    @SubscribeEvent
    fun onClientTick(event: LastClientTickEvent) {
        if (serverTick) return
        if (tick <= 0) {
            MinecraftForge.EVENT_BUS.unregister(this)
            block()
        }
        tick--
    }

    @Suppress("unused")
    @SubscribeEvent
    fun onServerTick(event: ServerTickEvent) {
        if (!serverTick) return
        if (tick <= 0) {
            MinecraftForge.EVENT_BUS.unregister(this)
            block()
        }
        tick--
    }
}